text/0000-async.md

```diff
@@ -0,0 +1,774 @@
+        self.spawn_obj(TaskObj::new(task))
+    }
+    // ...
+}
+```
+
+At that point we would also deprecate `TaskObj`, which is the reason for using
+the `Obj` suffix -- we want to keep the name `Task` available for potential
+usage down the line.
+
+In addition to the above, the `core::task` module will include the following API
+for helping detect bugs:
+
+```rust
+/// Marks the current thread as being within the dynamic extent of an
+/// executor.
+///
+/// Executor implementations should call this function before beginning to
+/// execute a tasks, and drop the returned `Enter` value after completing
+/// task execution:
+///
+/// ```rust
+/// let enter = enter().expect("...");
+/// /* run task */
+/// drop(enter);
+/// ```
+///
+/// Doing so ensures that executors aren't accidentally invoked in a nested fashion.
+/// When that happens, the inner executor can block waiting for an event that can
+/// only be triggered by the outer executor, leading to a deadlock.
+///
+/// # Error
+///
+/// Returns an error if the current thread is already marked, in which case the
+/// caller should panic with a tailored error message.
+pub fn enter() -> Result<Enter, EnterError>
+```
+
+As stated in the doc comment, the expectation is that all executors will wrap
+their task execution within an `enter` to detect inadvertent nesting.
+
+### Task contexts
+
+All tasks are executed with two pieces of contextual information:
+
+- A `Waker` for waking the task up later on.
+- An executor, which is the "default" place to spawn further tasks.
+
+Notably, this list does *not* include task-local data; that can be addressed
+externally, as we'll see in a later section.
+
+The `core::task::Context` type gathers the (stack-rooted) contextual information
+together, and is passed by mutable reference to all polling functions:
+
+```rust
+/// Information about the currently-running task.
+///
+/// Contexts are always tied to the stack, since they are set up specifically
+/// when performing a single `poll` step on a task.
+pub struct Context<'a> { .. }
+
+impl<'a> Context<'a> {
+    pub fn new(waker: &'a Waker, executor: &'a mut Executor) -> Context<'a>
+
+    /// Get the `Waker` associated with the current task.
+    pub fn waker(&self) -> &Waker;
+
+    /// Run an asynchronous computation to completion on the default executor.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if the default executor is unable to spawn.
+    /// To handle executor errors, use the `executor` method instead.
+    pub fn spawn(&mut self, f: impl Async<Output = ()> + 'static + Send);
+
+    /// Get the default executor associated with this task.
+    ///
+    /// This method is useful primarily if you want to explicitly handle
+    /// spawn failures.
+    ///
+    /// NB: this will remain unstable until the final `Executor` trait is ready.
+    pub fn executor(&mut self) -> &mut BoxExecutor;
```

Comment:
`BoxExecutor` is not defined anywhere. Is it possible to implement `FuturesUnordered` with this current definition?

Also, given the known issues* with this strategy, would it be wiser to punt on executor being tied to context?

\* spawning in drop and spawning from a function.

Reply:
This should be `Executor`.

We can always delay on stabilizing this portion of the API if we are feeling uneasy. Remember, everything that goes into std goes on the *nightly* channel indefinitely; there's a separate process for actually committing it to stable.